File: apps/pwa/lib/api/schedules.ts
/** Schedules API — REAL backend `/api/v1/schedules` (backend/api/v1/schedules.py):
 *  the authoring surface for `workspace_schedules`, the channel that lets BSVibe
 *  start work on its own. Before this the `ScheduleWorker` polled the table but
 *  nothing wrote rows — a dead channel; these are its producer.
 *
 *   POST   /api/v1/schedules        — author one schedule. Body is
 *                                      {@link ScheduleCreate}. The `text`
 *                                      instruction IS the scheduled run's task.
 *   GET    /api/v1/schedules        — list this workspace's schedules, newest first.
 *   DELETE /api/v1/schedules/{id}   — remove a schedule (204 No Content).
 *   PATCH  /api/v1/schedules/{id}   — enable / disable a schedule.
 *
 *  The backend accepts TWO kinds today — `instruction` and `product_tick`
 *  (`_SUPPORTED_KINDS` in `backend/schedule/application/schedule_service.py`;
 *  `product_tick` shipped 2026-07-21 in PR #609 and is live in prod). THIS UI
 *  authors `instruction` only. Read that as the UI's current scope, not as a
 *  claim about the capability: `product_tick` is authorable today through
 *  MCP/REST, and is absent here only because it takes no `text` and REQUIRES a
 *  `product_id` — a different form plus a product picker, a product decision
 *  still open in #948. `skill` / `plugin_action` genuinely do NOT exist (they
 *  are not in `_SUPPORTED_KINDS`). That split is machine-checked in
 *  `test/schedule-kinds-backend-parity.test.ts` so it cannot silently rot the
 *  way the sentence this one replaces did. Times (`next_run_at`) come back as
 *  UTC ISO strings; the surface formats them in the workspace time zone. */

import { apiFetch } from "./client";

/** The kind this UI authors — `payload={"text": <what to do>}`. NOT the only
 *  kind the backend accepts; see {@link BACKEND_SUPPORTED_SCHEDULE_KINDS}. */
export const SCHEDULE_KIND_INSTRUCTION = "instruction";

/** An autonomous cadence tick: the founder sets only WHEN (per product) and
 *  BSVibe decides WHAT at fire time. The backend accepts it; this UI does not
 *  author it yet (no product picker — #948), so it is created via MCP/REST. */
export const SCHEDULE_KIND_PRODUCT_TICK = "product_tick";

/** Every kind `POST /api/v1/schedules` accepts — a mirror of the backend's
 *  `_SUPPORTED_KINDS`, not an independent opinion about what exists. Pinned to
 *  the real frozenset by `test/schedule-kinds-backend-parity.test.ts`: adding a
 *  kind there without updating this goes red. */
export const BACKEND_SUPPORTED_SCHEDULE_KINDS = [
  SCHEDULE_KIND_INSTRUCTION,
  SCHEDULE_KIND_PRODUCT_TICK,
] as const;

/** Every kind THIS UI can author today — deliberately a SUBSET of the above.
 *  The gap is unbuilt UI, never an unbuilt capability; keeping the two lists
 *  apart is the whole of #948. */
export const PWA_AUTHORABLE_SCHEDULE_KINDS = [SCHEDULE_KIND_INSTRUCTION] as const;

/** A stored schedule row — mirrors the backend `ScheduleView`. `next_run_at` /
 *  `last_fired_at` are UTC ISO datetime strings. */
export interface Schedule {
  id: string;
  kind: string;
  text: string;
  cron_expr: string;
  product_id: string | null;
  title: string | null;
  next_run_at: string;
  last_fired_at: string | null;
  enabled: boolean;
}

/** Request body for authoring a schedule — mirrors the backend `ScheduleCreate`
 *  (`extra=forbid`). `kind` defaults to `instruction`, which is the only kind
 *  this UI sends today (see {@link PWA_AUTHORABLE_SCHEDULE_KINDS}); the field is
 *  open because the backend also takes `product_tick`. */
export interface ScheduleCreate {
  kind?: string;
  text: string;
  cron_expr: string;
  product_id?: string | null;
  title?: string | null;
}

/** List this workspace's schedules, newest first. */
export function getSchedules(): Promise<Schedule[]> {
  return apiFetch<Schedule[]>("/api/v1/schedules");
}

/** Author one schedule. 400 on an invalid cron expression / unsupported kind. */
export function createSchedule(body: ScheduleCreate): Promise<Schedule> {
  const payload: ScheduleCreate = {
    kind: body.kind ?? SCHEDULE_KIND_INSTRUCTION,
    text: body.text,
    cron_expr: body.cron_expr,
  };
  if (body.product_id != null) payload.product_id = body.product_id;
  if (body.title != null && body.title !== "") payload.title = body.title;
  return apiFetch<Schedule>("/api/v1/schedules", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/** Delete a schedule. 204 No Content, so this resolves to void. */
export function deleteSchedule(id: string): Promise<void> {
  return apiFetch<void>(`/api/v1/schedules/${id}`, { method: "DELETE" });
}

/** Enable or disable a schedule (PATCH). Returns the reconciled row. */
export function setScheduleEnabled(id: string, enabled: boolean): Promise<Schedule> {
  return apiFetch<Schedule>(`/api/v1/schedules/${id}`, {
    method: "PATCH",
    body: JSON.stringify({ enabled }),
  });
}

File: backend/api/v1/schedules.py
"""/api/v1/schedules — author the schedules that let BSVibe start work on its own.

The producer surface of the ``workspace_schedules`` channel (S1). Before this
the ``ScheduleWorker`` polled that table in prod but nothing wrote rows to it —
a dead channel. These endpoints are the authoring input:

* ``POST   /api/v1/schedules``       — create a natural-language ``instruction``
  schedule (``{kind, text, cron_expr, product_id?, title?}``). The instruction
  IS the run task; the emitter carries ``text`` into the framer so a scheduled
  run frames the instruction (not "Untitled run").
* ``GET    /api/v1/schedules``       — list this workspace's schedules.
* ``DELETE /api/v1/schedules/{id}``  — remove a schedule.
* ``PATCH  /api/v1/schedules/{id}``  — enable / disable a schedule.

Workspace resolution mirrors the sibling routers (:mod:`backend.api.v1.notifications`):
the ``get_workspace_id`` dep publishes the workspace into the ORM-scoping
contextvar + Postgres RLS GUC. S1 is the ``instruction`` kind only; other kinds
(skill / product_tick / plugin_action) are S4.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.ext.asyncio import AsyncSession

from backend.api.deps import get_db_session, get_workspace_id
from backend.schedule.application.schedule_service import (
    ScheduleService,
    ScheduleValidationError,
)
from backend.schedule.infrastructure.schedule_db import (
    SCHEDULE_KIND_INSTRUCTION,
    WorkspaceScheduleRow,
)

router = APIRouter()


class ScheduleCreate(BaseModel):
    """Request body for authoring a schedule (S1: ``instruction`` kind only)."""

    model_config = ConfigDict(extra="forbid")

    kind: str = Field(default=SCHEDULE_KIND_INSTRUCTION)
    text: str = Field(min_length=1, max_length=4000)
    cron_expr: str = Field(min_length=1, max_length=255)
    product_id: uuid.UUID | None = None
    title: str | None = Field(default=None, max_length=500)


class ScheduleEnabledPatch(BaseModel):
    """Request body for enabling / disabling a schedule."""

    model_config = ConfigDict(extra="forbid")

    enabled: bool


class ScheduleView(BaseModel):
    """Response shape for a schedule row."""

    model_config = ConfigDict(extra="forbid")

    id: uuid.UUID
    kind: str
    text: str
    cron_expr: str
    product_id: uuid.UUID | None
    title: str | None
    next_run_at: datetime
    last_fired_at: datetime | None
    enabled: bool


def _to_view(row: WorkspaceScheduleRow) -> ScheduleView:
    payload: dict[str, Any] = row.payload or {}
    text_value = payload.get("text")
    return ScheduleView(
        id=row.id,
        kind=row.kind,
        text=text_value if isinstance(text_value, str) else "",
        cron_expr=row.cron_expr,
        product_id=row.product_id,
        title=row.title,
        next_run_at=row.next_run_at,
        last_fired_at=row.last_fired_at,
        enabled=row.enabled,
    )


@router.post(
    "",
    status_code=status.HTTP_201_CREATED,
    response_model=ScheduleView,
    operation_id="create_schedule",
)
async def create_schedule(
    payload: ScheduleCreate,
    workspace_id: Annotated[uuid.UUID, Depends(get_workspace_id)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
) -> ScheduleView:
    """Author one schedule. 400 on an invalid cron expression / unsupported kind."""
    service = ScheduleService(session)
    try:
        row = await service.create(
            workspace_id=workspace_id,
            kind=payload.kind,
            text=payload.text,
            cron_expr=payload.cron_expr,
            product_id=payload.product_id,
            title=payload.title,
        )
    except ScheduleValidationError as exc:
        await session.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    await session.commit()
    return _to_view(row)


@router.get("", response_model=list[ScheduleView], operation_id="list_schedules")
async def list_schedules(
    workspace_id: Annotated[uuid.UUID, Depends(get_workspace_id)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
) -> list[ScheduleView]:
    """List this workspace's schedules, newest first."""
    service = ScheduleService(session)
    rows = await service.list(workspace_id=workspace_id)
    return [_to_view(row) for row in rows]


@router.delete(
    "/{schedule_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    operation_id="delete_schedule",
)
async def delete_schedule(
    schedule_id: uuid.UUID,
    workspace_id: Annotated[uuid.UUID, Depends(get_workspace_id)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
) -> None:
    """Delete a schedule. 404 if it does not exist in this workspace."""
    service = ScheduleService(session)
    deleted = await service.delete(schedule_id=schedule_id, workspace_id=workspace_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=f"schedule {schedule_id} not found"
        )
    await session.commit()


@router.patch(
    "/{schedule_id}",
    response_model=ScheduleView,
    operation_id="set_schedule_enabled",
)
async def set_schedule_enabled(
    schedule_id: uuid.UUID,
    payload: ScheduleEnabledPatch,
    workspace_id: Annotated[uuid.UUID, Depends(get_workspace_id)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
) -> ScheduleView:
    """Enable or disable a schedule. 404 if it does not exist in this workspace."""
    service = ScheduleService(session)
    row = await service.set_enabled(
        schedule_id=schedule_id, workspace_id=workspace_id, enabled=payload.enabled
    )
    if row is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=f"schedule {schedule_id} not found"
        )
    await session.commit()
    return _to_view(row)


__all__ = [
    "ScheduleCreate",
    "ScheduleEnabledPatch",
    "ScheduleView",
    "router",
]
